core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -461,7 +467,8 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
     // If replica failure did not require leader re-election, inform brokers of the offline replica
     // Note that during leader re-election, brokers update their metadata
     if (partitionsWithoutLeader.isEmpty) {
-      sendUpdateMetadataRequest(controllerContext.liveOrShuttingDownBrokerIds.toSeq)
+      sendUpdateMetadataRequest(controllerContext.liveOrShuttingDownBrokerIds.toSeq,
+        newOfflineReplicas.map(_.topicPartition).toSet)
```

Comment:
The purpose of sendUpdateMetadataRequest() here seems to be just propagating the live broker list since any partition level metadata would have been propagated through the partitionStateMachine and the replicaStateMachine logic above. The original code probably didn't realize that an empty partition list implies including all partitions' metadata. So, here, we probably just want to pass in an empty partition list.

Reply:
You are right. Previously I thought that the offline replicas info are not sent if `partitionsWithoutLeader` is empty but indeed in `replicaStateMachine.handleStateChanges(newOfflineReplicasNotForDeletion.toSeq, OfflineReplica)` we will send LeaderAndIsr and UpdateMetadata with offline replicas info to affected brokers.

Fixed.